pkg/reconciler/rolloutorchestrator/rolloutorchestrator.go

```diff
@@ -0,0 +1,379 @@
+			result = append(result, r)
+		}
+	}
+	return result
+}
+
+func targetRevisionEqual(currentStatusRevisions, finalTargetRevisions []v1.TargetRevision) bool {
+	if *finalTargetRevisions[0].Percent != 100 {
+		return false
+	}
+	for _, r := range currentStatusRevisions {
+		if *r.Percent == 100 && r.RevisionName == finalTargetRevisions[0].RevisionName {
+			return true
+		}
+	}
+	return false
+}
+
+// UpdateSPAForRevUp update the SPA(StagePodAutoscaler) for the revision scaling up, based on the TargetReplicas
+// min & max scales defined in the Knative Service.
+func UpdateSPAForRevUp(spa *v1.StagePodAutoscaler, revision *v1.TargetRevision, _ bool) *v1.StagePodAutoscaler {
+	// For revisions scaling up, the StageMaxScale is always set to the final MaxScale.
+	spa.Spec.StageMaxScale = revision.MaxScale
+	min := getMinScale(revision)
+
+	// TargetReplicas is nil, when either the revision has 0% traffic assigned or 100% traffic assigned.
+	// It will be either the old revision scales down to 0 or the new revision scale up to replace the old revision.
+	// In either case, the spa pick up the StageMinScale and StageMaxScale scales from the knative service.
+	if revision.TargetReplicas == nil {
+		spa.Spec.StageMinScale = revision.MinScale
+		return spa
+	}
+	targetReplicas := *revision.TargetReplicas
+	if targetReplicas < min && *revision.Percent < int64(100) {
+		// If the less than of 100% traffic is assigned to this revision or targetReplicas is less than minscale,
+		// set StageMinScale directly to targetReplicas.
+		spa.Spec.StageMinScale = ptr.Int32(targetReplicas)
+	} else {
+		// If the 100% traffic is assigned to this revision or targetReplicas is equal to or greater than min,
+		// set StageMinScale directly to the final MinScale from the knative service.
+		spa.Spec.StageMinScale = revision.MinScale
+	}
+	return spa
+}
+
+// UpdateSPAForRevDown update the SPA(StagePodAutoscaler) for the revision scaling down, based on the TargetReplicas
+// min & max scales defined in the Knative Service, if the scaleUpReady is true.
+//
+// If the scaleUpReady is false, no change to the SPA(StagePodAutoscaler).
+func UpdateSPAForRevDown(spa *v1.StagePodAutoscaler, revision *v1.TargetRevision,
+	scaleUpReady bool) *v1.StagePodAutoscaler {
+	if !scaleUpReady {
+		return spa
+	}
+	min := getMinScale(revision)
+	max := getMaxScale(revision)
+
+	if revision.TargetReplicas == nil {
+		spa.Spec.StageMinScale = revision.MinScale
+		spa.Spec.StageMaxScale = revision.MaxScale
+		return spa
+	}
+	targetReplicas := *revision.TargetReplicas
+
+	// If targetReplicas is equal to or greater than maxScale, StageMinScale and StageMaxScale are set to the final
+	// MinScale and MaxScale.
+	if targetReplicas >= max {
+		spa.Spec.StageMaxScale = revision.MaxScale
+		spa.Spec.StageMinScale = revision.MinScale
+		return spa
+	}
+
+	// If targetReplicas is less than maxScale, StageMaxScale is set to targetReplicas.
+	spa.Spec.StageMaxScale = ptr.Int32(targetReplicas)
```

Comment:
if targetReplicas >= max is true, this will override what is set in line 248 right?

Reply:
This function is to update the SPA for the revision to scale down.
If targetReplicas >= max is true, StageMaxScale is set to final MaxScale, and StageMinScale is set to MinScale, Line 250 return spa, Line 254 will not be called.

Line 254 is called, when targetReplicas < max. Since this is the revision to scale down, I need to use StageMaxScale to set the upper limit to make sure this revision will not go above targetReplicas.